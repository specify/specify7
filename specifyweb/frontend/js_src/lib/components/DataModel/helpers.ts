import { f } from '../../utils/functools';
import type { RA, ValueOf } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type {
  AnySchema,
  AnyTree,
  SerializedResource,
  TableFields,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import type { LiteralField, Relationship } from './specifyField';
import type { SpecifyTable } from './specifyTable';
import type { Tables } from './types';
/**
 * Lookups for operaters and relationships in Queries on the backend are
 * separated by `__`
 * The api supports the same syntax in query paramters.
 * For example `/api/specify/collectionobject/?collection__discipline__name__startswith="Invert"&catalognumber__gt=000000100`
 * fetches all collectionobjects in disciplines which names start with
 * "Invert" with catalognumbers greater than 100
 */
export const djangoLookupSeparator = '__';

export const backboneFieldSeparator = '.';

/**
 * Formats a relationship lookup which is used in a query passed to the backend.
 * For example `formatRelationshipPath('collection', 'discipline', 'division')`
 * becomes `'collection__discipline__division'`
 */
export const formatRelationshipPath = (...fields: RA<string>): string =>
  fields.join(djangoLookupSeparator);

const weekDayMap = {
  Sunday: 1,
  Monday: 2,
  Tuesday: 3,
  Wednesday: 4,
  Thursday: 5,
  Friday: 6,
  Saturday: 7,
};

/**
 * Use this to construct a query using a lookup for Django.
 * Returns an object which can be used as a filter when fetched from the backend.
 * Example: backendFilter('number1').isIn([1, 2, 3]) is the equivalent
 * of {number1__in: [1, 2, 3].join(',')}
 *
 * See the Django docs at:
 * https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups
 */
export const backendFilter = (field: string) => ({
  equals: (value: number | string) => ({
    [[field, 'exact'].join(djangoLookupSeparator)]: value,
  }),
  contains: (value: string) => ({
    [[field, 'contains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveContains: (value: string) => ({
    [[field, 'icontains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveStartsWith: (value: string) => ({
    [[field, 'istartswith'].join(djangoLookupSeparator)]: value,
  }),
  startsWith: (value: string) => ({
    [[field, 'startswith'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveEndsWith: (value: string) => ({
    [[field, 'iendswith'].join(djangoLookupSeparator)]: value,
  }),
  endsWith: (value: string) => ({
    [[field, 'endswith'].join(djangoLookupSeparator)]: value,
  }),
  isIn: (value: RA<number | string>) => ({
    [[field, 'in'].join(djangoLookupSeparator)]: value.join(','),
  }),
  isNull: (value: 'false' | 'true' = 'true') => ({
    [[field, 'isnull'].join(djangoLookupSeparator)]: value,
  }),
  greaterThan: (value: number) => ({
    [[field, 'gt'].join(djangoLookupSeparator)]: value,
  }),
  greaterThanOrEqualTo: (value: number) => ({
    [[field, 'gte'].join(djangoLookupSeparator)]: value,
  }),
  lessThan: (value: number) => ({
    [[field, 'lt'].join(djangoLookupSeparator)]: value,
  }),
  lessThanOrEqualTo: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  matchesRegex: (value: string) => ({
    [[field, 'regex'].join(djangoLookupSeparator)]: value,
  }),

  dayEquals: (value: number) => ({
    [[field, 'day'].join(djangoLookupSeparator)]: value,
  }),
  monthEquals: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  yearEquals: (value: number) => ({
    [[field, 'year'].join(djangoLookupSeparator)]: value,
  }),
  weekEquals: (value: number) => ({
    [[field, 'week'].join(djangoLookupSeparator)]: value,
  }),
  weekDayEquals: (value: keyof typeof weekDayMap) => ({
    [[field, 'week_day'].join(djangoLookupSeparator)]: weekDayMap[value],
  }),
});

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyTable.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyTable.name === tableName
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * The table.field has a very broad type to reduce type conflicts in components
 * that deal with generic schemas (accept AnySchema or a SCHEMA extends AnySchema)
 */
export const getField = <
  SCHEMA extends ValueOf<Tables>,
  FIELD extends TableFields<SCHEMA>
>(
  table: SpecifyTable<SCHEMA>,
  name: FIELD
): FIELD extends keyof SCHEMA['fields'] ? LiteralField : Relationship =>
  table.field[name] as FIELD extends keyof SCHEMA['fields']
    ? LiteralField
    : Relationship;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyTable.name)
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * Example usage:
 * resource: Collector
 * fields: agent -> lastName
 * Would return [agent, lastName] if agent exists
 *
 */
export async function fetchDistantRelated(
  resource: SpecifyResource<AnySchema>,
  fields: RA<LiteralField | Relationship> | undefined
): Promise<
  | {
      readonly resource: SpecifyResource<AnySchema> | undefined;
      readonly field: LiteralField | Relationship | undefined;
    }
  | undefined
> {
  if (
    Array.isArray(fields) &&
    fields.some(
      (field) =>
        field.isRelationship &&
        relationshipIsToMany(field) &&
        field !== fields.at(-1)
    )
  ) {
    console.error(
      'Can not index inside of a -to-many relationship. Use an aggregator instead'
    );
    return undefined;
  }

  const related =
    fields === undefined || fields.length === 0
      ? resource
      : fields.length === 1
      ? await resource.fetch()
      : await resource.rgetPromise(
          fields
            .slice(0, -1)
            .map(({ name }) => name)
            .join(backboneFieldSeparator)
        );

  const field = fields?.at(-1);
  const relatedResource = related ?? undefined;
  return relatedResource === undefined && field === undefined
    ? undefined
    : {
        resource: relatedResource,
        field,
      };
}
