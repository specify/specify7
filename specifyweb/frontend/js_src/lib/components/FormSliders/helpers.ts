import { Relationship } from '../DataModel/specifyField';

export const shouldBeToOne = (
  relationship: Relationship | undefined
): boolean =>
  relationship?.type.includes('-to-many') === false ||
  (relationship?.type.includes('-to-many') === true &&
    !relationship.isDependent() &&
    relationship.getReverse()?.isDependent() === true);
