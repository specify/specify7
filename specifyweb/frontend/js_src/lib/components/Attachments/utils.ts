import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
