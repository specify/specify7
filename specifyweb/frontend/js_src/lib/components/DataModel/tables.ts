/**
 * Fetch Specify DataModel JSON and reify it into the SpecifyTable and
 * SpecifyField objects
 */

import type { LocalizedString } from 'typesafe-i18n';

import { f } from '../../utils/functools';
import type { IR, RA, RR } from '../../utils/types';
import {
  defined,
  overwriteReadOnly,
  setDevelopmentGlobal,
} from '../../utils/types';
import { sortFunction, split } from '../../utils/utils';
import { error } from '../Errors/assert';
import { load } from '../InitialContext';
import { isTreeTable } from '../InitialContext/treeRanks';
import { formatUrl } from '../Router/queryString';
import type { AnySchema, AnyTree } from './helperTypes';
import { schemaExtras } from './schemaExtras';
import { LiteralField, Relationship } from './specifyField';
import { type TableDefinition, SpecifyTable } from './specifyTable';
import type { Agent, Tables } from './types';

export type SchemaLocalization = {
  readonly name: LocalizedString | null;
  readonly desc: LocalizedString | null;
  readonly format: string | null;
  readonly aggregator: string | null;
  readonly ishidden: boolean;
  readonly items: IR<{
    readonly name: LocalizedString | null;
    readonly desc: LocalizedString | null;
    readonly format: string | null;
    readonly picklistname: string | null;
    readonly weblinkname: string | null;
    readonly isrequired: boolean;
    readonly ishidden: boolean;
  }>;
};

const processFields = <FIELD_TYPE extends LiteralField | Relationship>(
  fields: RA<FIELD_TYPE>,
  frontEndFields: RA<FIELD_TYPE>
): RA<FIELD_TYPE> =>
  [
    ...fields,
    ...frontEndFields.map((field) => {
      field.overrides.isReadOnly = true;
      field.isVirtual = true;
      return field;
    }),
    /*
     * The sort order defined here affects the order of fields in the
     * WbPlanView, Query builder, Schema Config and all other places
     */
  ].sort(sortFunction(({ label = '' }) => label));

let schemaLocalization: IR<SchemaLocalization> = undefined!;
const fetchSchemaLocalization = f.store(async () =>
  import('../Preferences/userPreferences').then(async ({ userPreferences }) =>
    load<IR<SchemaLocalization>>(
      formatUrl('/context/schema_localization.json', {
        lang: userPreferences.get('form', 'schema', 'language'),
      }),
      'application/json'
    )
  )
);
export const getSchemaLocalization = (): IR<SchemaLocalization> =>
  schemaLocalization ??
  error('Accessing schema localization before fetching it');

const frontEndOnlyFields: Partial<Record<keyof Tables, RA<string>>> = {};
export const getFrontEndOnlyFields = (): Partial<
  RR<keyof Tables, RA<string>>
> => frontEndOnlyFields;

export const tables = {} as {
  readonly [TABLE_NAME in keyof Tables]: SpecifyTable<Tables[TABLE_NAME]>;
};
export const genericTables = tables as RR<keyof Tables, SpecifyTable>;

export const fetchContext = f
  .all({
    dataModel: load<RA<TableDefinition>>(
      '/context/datamodel.json',
      'application/json'
    ),
    localization: fetchSchemaLocalization(),
  })
  .then(({ dataModel, localization }) => {
    schemaLocalization = localization;
    dataModel
      .map((tableDefinition) => {
        const table = new SpecifyTable(tableDefinition);
        overwriteReadOnly(genericTables, table.name, table);
        return [tableDefinition, table] as const;
      })
      .forEach(([tableDefinition, table]) => {
        const [frontEndFields, callback] = (
          schemaExtras[table.name] as typeof schemaExtras['Agent'] | undefined
        )?.(table as SpecifyTable<Agent>) ?? [[]];
        const [literalFields, relationships] = split(
          frontEndFields.map((field) => {
            field.isReadOnly = true;
            field.overrides.isReadOnly = true;
            return field;
          }),
          (field) => field.isRelationship
        );

        overwriteReadOnly(
          table,
          'literalFields',
          processFields(
            tableDefinition.fields.map(
              (fieldDefinition) => new LiteralField(table, fieldDefinition)
            ),
            literalFields
          )
        );
        overwriteReadOnly(
          table,
          'relationships',
          processFields(
            tableDefinition.relationships.map(
              (relationshipDefinition) =>
                new Relationship(table, relationshipDefinition)
            ),
            relationships
          )
        );
        overwriteReadOnly(table, 'fields', [
          ...table.literalFields,
          ...table.relationships,
        ]);
        overwriteReadOnly(
          table,
          'field',
          Object.fromEntries(table.fields.map((field) => [field.name, field]))
        );

        frontEndOnlyFields[table.name] = [
          ...literalFields,
          ...relationships,
        ].map(({ name }) => name);

        callback?.();
      });
    return tables;
  });

setDevelopmentGlobal('_tables', genericTables);

/**
 * Returns a schema table object describing the named Specify model
 * Can wrap this function call in defined() to cast result to SpecifyTable
 *
 * @remarks
 * This function is useful when getting table using a dynamic name. If name is
 * known statically, can get table directly on the "tables" object
 */
export function getTable(name: string): SpecifyTable | undefined {
  if (
    process.env.NODE_ENV !== 'production' &&
    Object.keys(genericTables).length === 0
  )
    // eslint-disable-next-line functional/no-throw-statement
    throw new Error(
      `Calling getTable() before data model is fetched.${
        process.env.NODE_ENV === 'test'
          ? ' If this is part of a test, remember to include requireContext() at the top of the test file'
          : ''
      }`
    );

  const lowerCase = name.toLowerCase();
  return name === ''
    ? undefined
    : genericTables[name as keyof Tables] ??
        Object.values(genericTables).find(
          (table) => table.name.toLowerCase() === lowerCase
        ) ??
        Object.values(genericTables).find(
          (table) => table.longName.toLowerCase() === lowerCase
        );
}

export const strictGetTable = (name: string): SpecifyTable =>
  defined(getTable(name), `Trying to get unknown table: ${name}`);

export function getTreeTable(name: string): SpecifyTable<AnyTree> | undefined {
  const table = getTable(name);
  if (typeof table === 'object' && !isTreeTable(table.name))
    // eslint-disable-next-line functional/no-throw-statement
    throw new Error(`${name} is not a tree table`);
  return table as unknown as SpecifyTable<AnyTree> | undefined;
}

/**
 * Looks up a schema table object describing Specify model using the Specify
 * tableId integer
 */
export const getTableById = <SCHEMA extends AnySchema>(
  tableId: number
): SpecifyTable<SCHEMA> =>
  (Object.values(genericTables).find((table) => table.tableId === tableId) as
    | SpecifyTable<SCHEMA>
    | undefined) ?? error(`Table with id ${tableId} does not exist`);

// If this is true, then you can use {domainfilter:true} when fetching that table
export const hasHierarchyField = (table: SpecifyTable): boolean =>
  [
    'collectionObject',
    'collection',
    'discipline',
    'division',
    'institution',
  ].some((fieldName) =>
    table.relationships.some(({ name }) => name === fieldName)
  );
