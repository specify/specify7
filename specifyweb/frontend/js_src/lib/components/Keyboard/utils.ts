import type { LocalizedString } from 'typesafe-i18n';

import type { RA } from '../../utils/types';
import { localized } from '../../utils/types';
import type { KeyboardShortcuts, ModifierKey } from './config';
import { keyboardModifierLocalization, keyboardPlatform } from './config';
import { keyJoinSymbol } from './context';

const localizedKeyJoinSymbol = ' + ';
export const localizeKeyboardShortcut = (shortcut: string): LocalizedString =>
  localized(
    shortcut
      .split(keyJoinSymbol)
      .map((key) => keyboardModifierLocalization[key as ModifierKey] ?? key)
      .join(localizedKeyJoinSymbol)
  );

/**
 * If there is a keyboard shortcut defined for current system, use it
 * (also, if current system explicitly has empty array of shortcuts, use it).
 *
 * Otherwise, use the keyboard shortcut from one of the other platforms if set,
 * but change meta to ctrl and vice versa as necessary.
 */
export function resolvePlatformShortcuts(
  shortcut: KeyboardShortcuts
): RA<string> | undefined {
  if ('platform' in shortcut) return shortcut[keyboardPlatform];
  else if ('other' in shortcut)
    return keyboardPlatform === 'windows'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('windows' in shortcut)
    return keyboardPlatform === 'other'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('mac' in shortcut) return shortcut.other?.map(replaceMetaWithCtrl);
  else return undefined;
}

const replaceCtrlWithMeta = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'ctrl' ? 'meta' : key))
    .join(keyJoinSymbol);

const replaceMetaWithCtrl = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'meta' ? 'ctrl' : key))
    .join(keyJoinSymbol);
