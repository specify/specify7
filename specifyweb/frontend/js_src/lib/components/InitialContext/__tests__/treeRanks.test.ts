import { requireContext } from '../../../tests/helpers';
import { theories } from '../../../tests/utils';
import { tables } from '../../DataModel/tables';
import {
  allTrees,
  exportsForTests,
  getDisciplineTrees,
  getTreeDefinitionItems,
  isTreeResource,
  isTreeTable,
  strictGetTreeDefinitionItems,
} from '../treeRanks';

const { getTreeScope } = exportsForTests;

requireContext();

test('getDisciplineTrees', () =>
  expect(getDisciplineTrees()).toMatchSnapshot());

theories(isTreeTable, [
  { in: ['Taxon'], out: true },
  {
    name: 'GeologicTimePeriod is a tree model, even though we are not in a paleo discipline',
    in: ['GeologicTimePeriod'],
    out: true,
  },
  { in: ['Locality'], out: false },
]);

describe('isTreeResource', () => {
  test('Taxon', () =>
    expect(isTreeResource(new tables.Taxon.Resource())).toBe(true));
  test('GeologicTimePeriod is a tree resource, even though we are not in a paleo discipline', () =>
    expect(isTreeResource(new tables.GeologicTimePeriod.Resource())).toBe(
      true
    ));
  test('Locality', () =>
    expect(isTreeResource(new tables.Locality.Resource())).toBe(false));
});

describe('Get tree definition', () => {
  test('for the Taxon tree', () =>
    expect(getTreeDefinitionItems('Taxon', false)).toMatchSnapshot());
  test('for the Geography tree', () =>
    expect(getTreeDefinitionItems('Geography', true)).toMatchSnapshot());
  test('for a non-tree table', () =>
    expect(
      getTreeDefinitionItems('Locality' as 'Taxon', true)
    ).toBeUndefined());
  test('for a non-existent table', () =>
    expect(
      getTreeDefinitionItems('SomeTable' as 'Taxon', true)
    ).toBeUndefined());
});

describe('strictGetTreeDefinitionItems', () => {
  test('Taxon', () =>
    expect(strictGetTreeDefinitionItems('Taxon', false)).toEqual(
      getTreeDefinitionItems('Taxon', false)
    ));
  test('Locality', () =>
    expect(() =>
      strictGetTreeDefinitionItems('Locality' as 'Taxon', true)
    ).toThrow(/Unable to get tree ranks for a/u));
});

test('getTreeScope', () =>
  expect(Object.fromEntries(allTrees.map((tree) => [tree, getTreeScope(tree)])))
    .toMatchInlineSnapshot(`
      {
        "Geography": "discipline",
        "GeologicTimePeriod": "discipline",
        "LithoStrat": "discipline",
        "Storage": "institution",
        "Taxon": "discipline",
      }
    `));
