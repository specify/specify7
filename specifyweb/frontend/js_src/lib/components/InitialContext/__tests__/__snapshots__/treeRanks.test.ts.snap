// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get tree definition for the Geography tree 1`] = `
[
  {
    "_tableName": "GeographyTreeDefItem",
    "children": "/api/specify/geographytreedefitem/?parent=1",
    "createdByAgent": null,
    "fullNameSeparator": ", ",
    "id": 1,
    "isEnforced": true,
    "isInFullName": null,
    "modifiedByAgent": null,
    "name": "Planet",
    "parent": null,
    "rankId": 0,
    "remarks": null,
    "resource_uri": "/api/specify/geographytreedefitem/1/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:04",
    "timestampModified": "2012-08-09T12:30:04",
    "title": null,
    "treeDef": "/api/specify/geographytreedef/1/",
    "treeEntries": "/api/specify/geography/?definitionitem=1",
    "version": 0,
  },
  {
    "_tableName": "GeographyTreeDefItem",
    "children": "/api/specify/geographytreedefitem/?parent=3",
    "createdByAgent": null,
    "fullNameSeparator": ", ",
    "id": 3,
    "isEnforced": null,
    "isInFullName": true,
    "modifiedByAgent": null,
    "name": "Continent",
    "parent": "/api/specify/geographytreedefitem/1/",
    "rankId": 100,
    "remarks": null,
    "resource_uri": "/api/specify/geographytreedefitem/3/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:04",
    "timestampModified": "2012-08-09T12:30:04",
    "title": "Continent",
    "treeDef": "/api/specify/geographytreedef/1/",
    "treeEntries": "/api/specify/geography/?definitionitem=3",
    "version": 1,
  },
  {
    "_tableName": "GeographyTreeDefItem",
    "children": "/api/specify/geographytreedefitem/?parent=2",
    "createdByAgent": null,
    "fullNameSeparator": ", ",
    "id": 2,
    "isEnforced": null,
    "isInFullName": true,
    "modifiedByAgent": null,
    "name": "Country",
    "parent": "/api/specify/geographytreedefitem/3/",
    "rankId": 200,
    "remarks": null,
    "resource_uri": "/api/specify/geographytreedefitem/2/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:04",
    "timestampModified": "2012-08-09T12:30:04",
    "title": null,
    "treeDef": "/api/specify/geographytreedef/1/",
    "treeEntries": "/api/specify/geography/?definitionitem=2",
    "version": 1,
  },
  {
    "_tableName": "GeographyTreeDefItem",
    "children": "/api/specify/geographytreedefitem/?parent=4",
    "createdByAgent": null,
    "fullNameSeparator": ", ",
    "id": 4,
    "isEnforced": null,
    "isInFullName": true,
    "modifiedByAgent": null,
    "name": "State",
    "parent": "/api/specify/geographytreedefitem/2/",
    "rankId": 300,
    "remarks": null,
    "resource_uri": "/api/specify/geographytreedefitem/4/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:04",
    "timestampModified": "2012-08-09T12:30:04",
    "title": null,
    "treeDef": "/api/specify/geographytreedef/1/",
    "treeEntries": "/api/specify/geography/?definitionitem=4",
    "version": 0,
  },
  {
    "_tableName": "GeographyTreeDefItem",
    "children": "/api/specify/geographytreedefitem/?parent=5",
    "createdByAgent": null,
    "fullNameSeparator": ", ",
    "id": 5,
    "isEnforced": null,
    "isInFullName": true,
    "modifiedByAgent": null,
    "name": "County",
    "parent": "/api/specify/geographytreedefitem/4/",
    "rankId": 400,
    "remarks": null,
    "resource_uri": "/api/specify/geographytreedefitem/5/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:04",
    "timestampModified": "2012-08-09T12:30:04",
    "title": null,
    "treeDef": "/api/specify/geographytreedef/1/",
    "treeEntries": "/api/specify/geography/?definitionitem=5",
    "version": 0,
  },
]
`;

exports[`Get tree definition for the Taxon tree 1`] = `
[
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=21",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 21,
    "isEnforced": false,
    "isInFullName": false,
    "modifiedByAgent": null,
    "name": "Kingdom",
    "parent": "/api/specify/taxontreedefitem/5/",
    "rankId": 10,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/21/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=21",
    "version": 2,
  },
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=10",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 10,
    "isEnforced": false,
    "isInFullName": false,
    "modifiedByAgent": null,
    "name": "Phylum",
    "parent": "/api/specify/taxontreedefitem/21/",
    "rankId": 30,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/10/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=10",
    "version": 5,
  },
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=12",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 12,
    "isEnforced": true,
    "isInFullName": false,
    "modifiedByAgent": null,
    "name": "Class",
    "parent": "/api/specify/taxontreedefitem/10/",
    "rankId": 60,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/12/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=12",
    "version": 6,
  },
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=4",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 4,
    "isEnforced": true,
    "isInFullName": false,
    "modifiedByAgent": null,
    "name": "Order",
    "parent": "/api/specify/taxontreedefitem/12/",
    "rankId": 100,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/4/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=4",
    "version": 6,
  },
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=19",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 19,
    "isEnforced": true,
    "isInFullName": false,
    "modifiedByAgent": null,
    "name": "Family",
    "parent": "/api/specify/taxontreedefitem/4/",
    "rankId": 140,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/19/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=19",
    "version": 2,
  },
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=20",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 20,
    "isEnforced": false,
    "isInFullName": false,
    "modifiedByAgent": null,
    "name": "Subfamily",
    "parent": "/api/specify/taxontreedefitem/19/",
    "rankId": 150,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/20/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=20",
    "version": 3,
  },
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=9",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 9,
    "isEnforced": true,
    "isInFullName": true,
    "modifiedByAgent": null,
    "name": "Genus",
    "parent": "/api/specify/taxontreedefitem/20/",
    "rankId": 180,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/9/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=9",
    "version": 3,
  },
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=15",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 15,
    "isEnforced": false,
    "isInFullName": true,
    "modifiedByAgent": null,
    "name": "Subgenus",
    "parent": "/api/specify/taxontreedefitem/9/",
    "rankId": 190,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/15/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=15",
    "version": 1,
  },
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=2",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 2,
    "isEnforced": false,
    "isInFullName": true,
    "modifiedByAgent": null,
    "name": "Species",
    "parent": "/api/specify/taxontreedefitem/15/",
    "rankId": 220,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/2/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=2",
    "version": 2,
  },
  {
    "_tableName": "TaxonTreeDefItem",
    "children": "/api/specify/taxontreedefitem/?parent=22",
    "createdByAgent": null,
    "formatToken": null,
    "fullNameSeparator": " ",
    "id": 22,
    "isEnforced": false,
    "isInFullName": true,
    "modifiedByAgent": null,
    "name": "Subspecies",
    "parent": "/api/specify/taxontreedefitem/2/",
    "rankId": 230,
    "remarks": null,
    "resource_uri": "/api/specify/taxontreedefitem/22/",
    "textAfter": null,
    "textBefore": null,
    "timestampCreated": "2012-08-09T12:30:07",
    "timestampModified": "2012-08-09T12:30:07",
    "title": null,
    "treeDef": "/api/specify/taxontreedef/1/",
    "treeEntries": "/api/specify/taxon/?definitionitem=22",
    "version": 1,
  },
]
`;

exports[`getDisciplineTrees 1`] = `
[
  "Geography",
  "Storage",
  "Taxon",
]
`;
